Cleanup

andreyk0 opened the pull request.

andreyk0 commented:
@amarrella wanted to discuss the Handler API, I'm not loving the `handle` function :)
I think `m ()` makes sense if we change it to run `forever` (i.e. handle loop as opposed to single event), otherwise it should probably propagate error results to caller somehow.

amarrella commented on core/src/AWS/Lambda/Handler.hs:
```diff
@@ -1,31 +1,51 @@
-  catchError result (postError eventID . Error "Error" . encodeOutput)
+
+-- | Handle single event with a given runtime client and a callback
+handle
```
I'm not even sure we should expose `handle` to be honest. `handler` should be the only API used by apps. This is a runtime, not a simple function, and the runtime does not know how long it will be alive (AWS can send 1 or more events to a runtime, depending on how long the function is being kept "hot" in the backend), and this will crash if it only handled 1 event. In the examples we only use `handler` for this reason and it was a mistake on my part exposing `handle` as well, as this leads to an incorrect implementation.

andreyk0 replied:
Yeah, I was wondering why we expose both, thought it could be so that you'd have a chance to customize runtime somehow, e.g. tweak http client settings or some such.
If it's not needed then yeah - one function is better and I'd actually like to rename it to `handle` then.

amarrella replied:
I used `handler` because it seems pretty standard (other runtimes like https://github.com/Nike-Inc/hal and https://theam.github.io/aws-lambda-haskell-runtime/ use the same term). I don't think it makes much sense to customize the current client (we don't expose anything special) but if so, i would rename `handle` to something like `handlerWIthClient` (find a better name :P )

andreyk0 commented:
@amarrella  take a look, I've also updated JSON encoding as I've moved events around.
They were missing `toEncoding` implementations, started to add them but it was noisy (and we were exposing `modify` functions as part of the module interface), so, I've pushed that into a corner.

amarrella commented:
@andreyk0 looks fantastic, thank you! I see some issue in running `make sam-tests` can you investigate? It seems to be complaining about `GLIBC_2.27

andreyk0 commented:
> @andreyk0 looks fantastic, thank you! I see some issue in running `make sam-tests` can you investigate? It seems to be complaining about `GLIBC_2.27

Oh, thanks for trying, I'll investigate. This is likely due to Lambda runtime being on a much older glibc. I'll see if there's a newer Lambda runtime or whatever else can be done.

andreyk0 commented:
@amarrella pushed GW V2 models as well as it adds a bit of config to include data files.
`sam` test run is failing atm and that's because our unit tests are misleading, they test symmetry between encoding/decoding but that's just `Aeson`, which already works. We need to test ability to parse real JSON. I'll add that.

andreyk0 commented:
@amarrella Ok, pls try again. As they say "it works on my machine".
Turns out JSON was Ok (but it weren't obvious, like I said tests were meaningless and runtime err reporting was lacking a bit).
Improved tests, added `sam` paramters in the Makefile to actually read event data from STDIN.

andreyk0 commented:
Oh, and glibc should be aligned now with runtime as code is built in the "official" lambda build container. 
This works with stack on linux (I'm on linux) because it's able to mount itself from the host file system into the container, similiar to ghc, cache dirs and the rest. 
I've noticed the way it does it is by downloading itself into `~/.stack/..` dir (which is mounted into container for other reasons and works already). So, this *should* work on MacOS as well as far as I can tell but IDK for sure.

andreyk0 commented:
BTW we could thin down the size by flipping `lens` (huge) dependency to `microlens`. This however would require removing `wreq` as well as it too depends on the `lens`, perhaps some other time.

amarrella commented:
@andreyk0 great! I'm testing it now on macOs but as you said it should work. Nitpick: any reason why we are using ghc 8.8.2 instead of 8.8.3?  (at least this is what is seems from the build logs).

> BTW we could thin down the size by flipping lens (huge) dependency to microlens. This however would require removing wreq as well as it too depends on the lens, perhaps some other time.

About `lens`, yeah i thought about it but never did it because of the transitive dependency you mentioned, if we manage to get rid of `wreq` though i'm all for it! I think we can also drop `dhall` (and its instances) since it was more of an exercise than a really useful things to provide to consumer. We can move the instances in the example executable so we keep the library smaller.

andreyk0 commented:
> @andreyk0 great! I'm testing it now on macOs but as you said it should work. Nitpick: any reason why we are using ghc 8.8.2 instead of 8.8.3? (at least this is what is seems from the build logs).

Oh, not really, that was the version when I started working on wai stuff but I see there's a newer one now, will update.

> > BTW we could thin down the size by flipping lens (huge) dependency to microlens. This however would require removing wreq as well as it too depends on the lens, perhaps some other time.
> 
> About `lens`, yeah i thought about it but never did it because of the transitive dependency you mentioned, if we manage to get rid of `wreq` though i'm all for it! I think we can also drop `dhall` (and its instances) since it was more of an exercise than a really useful things to provide to consumer. We can move the instances in the example executable so we keep the library smaller.

I think having multi-module build should let us manage dependencies properly. That's why I set it up for wai. Dhall could work the same way (i.e. we could move it into a submodule) to minimize dependency requirements for projects that don't care.

wreq wouldn't be too hard to get rid of, it's a thin "lens" wrapper around `http-client`, nothing there that's not in `http-client`. I tried to do it but the change seemed larger than I anticipated and I reverted back to continue with my more immediate goals.

andreyk0 commented:
@amarrella bumped stack, works for me but I've noticed that the compilation is no longer `static`, this is because I split project into sub-modules and the parameter name should be different now:

```diff
diff --git a/Makefile b/Makefile
index 7cd19fe..9800a0e 100644
--- a/Makefile
+++ b/Makefile
@@ -3,7 +3,12 @@ EXAMPLES  = api-gateway kinesis s3 sns dhall
 
 .PHONY: build
 build: ## compile and test
-	stack build --flag aws-lambda-haskell-runtime-client:static --docker --test --copy-bins --local-bin-path ${BUILD_DIR}
+	stack build \
+    --flag aws-lambda-haskell-runtime-client-examples:static \
+    --docker \
+    --test \
+    --copy-bins \
+    --local-bin-path ${BUILD_DIR}
 
 .PHONY: dev
 dev: ## automatically compile and test when a file is written

```

However, if I fix that - it no longer compiles as Lambda environment doesn't have the necessary packages installed. I can fix that by wrapping `lambci` container into a customized version but I wonder if we should just try as is. 
Static linking with `glibc` is misleading at best, it only works so long as your statically linked version roughly matches what's on the system anyway. See https://stackoverflow.com/questions/11693320/linking-glibc-statically-and-proprietary-software-licensing

OTOH if we build in a "Lambda environment" anyway (that's what the switch to `lambci` container for build was all about) then same .so libraries should be present at runtime and linking them statically (even if we fix that) will only increase the size of the resulting binary (and a lambda cold start time) for no real benefit.

I propose we try to default to dynamic linking and remove references to static linking for now.

Note that if additional (not present in the default runtime) libraries are needed by projects using this lib they are probably better off using https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html and use a custom (derived from `lambci`)  docker build image that make said additional libraries available during build.

amarrella commented:
@andreyk0 makes sense and i'm ok with dynamic linking as long as we match the lambda environment 👍

andreyk0 commented:
@amarrella K, let me remove references to static compilation for now and maybe we can merge?
